File: Dome/dome/rules/rce.py
"""Remote Code / Command Execution detection rules."""
import re

PATTERNS = [
    # Shell metacharacters in param values
    (r"[;&|`$]\s*(ls|cat|pwd|id|whoami|uname|curl|wget|bash|sh|python|perl|ruby|nc|ncat)\b",
                                                                     "RCE-001", "Shell command in parameter"),
    # Backtick / $() command substitution
    (r"`[^`]+`|\$\([^)]+\)",                                         "RCE-002", "Command substitution"),
    # Pipe to shell
    (r"\|\s*(bash|sh|zsh|ksh|csh)\b",                                "RCE-003", "Pipe to shell"),
    # eval with encoded content
    (r"(?i)\beval\s*\(",                                             "RCE-004", "eval() call"),
    # OS command functions (PHP/Python/etc)
    (r"(?i)\b(system|passthru|popen|proc_open|shell_exec|exec|subprocess)\s*\(", "RCE-005", "OS command function"),
    # Base64-encoded payloads piped to bash
    (r"(?i)base64\s*(-d|--decode).*?\|\s*(bash|sh)",                 "RCE-006", "Base64-decoded shell execution"),
    # curl/wget piped to shell (dropper pattern)
    (r"(?i)(curl|wget)[^|]+\|\s*(bash|sh|python)",                   "RCE-007", "Download and execute dropper"),
    # /dev/tcp reverse shell
    (r"/dev/tcp/",                                                   "RCE-008", "Bash /dev/tcp reverse shell"),
    # OGNL injection (Struts/Java)
    (r"(?i)%\{[^}]*\}|#\{[^}]*\}",                                  "RCE-009", "OGNL/EL expression injection"),
    # Log4Shell / JNDI
    (r"(?i)\$\{jndi:",                                               "RCE-010", "Log4Shell JNDI injection"),
]

COMPILED = [(re.compile(p), rid, desc) for p, rid, desc in PATTERNS]


def check(value: str) -> list[dict]:
    hits = []
    for pattern, rule_id, description in COMPILED:
        if pattern.search(value):
            hits.append({"rule_id": rule_id, "description": description, "category": "rce"})
    return hits

File: Dome/dome-rs/src/engine/rce.rs
/// Remote Code / Command Execution detection.
use once_cell::sync::Lazy;
use regex::Regex;
use super::Hit;

const SCORE: u32 = 35;

static RULES: Lazy<Vec<(Regex, &'static str, &'static str)>> = Lazy::new(|| {
    vec![
        (Regex::new(r"[;&|`$]\s*(ls|cat|pwd|id|whoami|uname|curl|wget|bash|sh|python|perl|ruby|nc|ncat)\b").unwrap(),
         "RCE-001", "Shell command in parameter"),
        (Regex::new(r"`[^`]+`|\$\([^)]+\)").unwrap(),
         "RCE-002", "Command substitution"),
        (Regex::new(r"\|\s*(bash|sh|zsh|ksh|csh)\b").unwrap(),
         "RCE-003", "Pipe to shell"),
        (Regex::new(r"\beval\s*\(").unwrap(),
         "RCE-004", "eval() call"),
        (Regex::new(r"\b(system|passthru|popen|proc_open|shell_exec|exec|subprocess)\s*\(").unwrap(),
         "RCE-005", "OS command function"),
        (Regex::new(r"base64\s*(-d|--decode).*\|\s*(bash|sh)").unwrap(),
         "RCE-006", "Base64-decoded shell execution"),
        (Regex::new(r"(curl|wget)[^|]+\|\s*(bash|sh|python)").unwrap(),
         "RCE-007", "Download and execute dropper"),
        (Regex::new(r"/dev/tcp/").unwrap(),
         "RCE-008", "Bash /dev/tcp reverse shell"),
        (Regex::new(r"%\{[^}]*\}|#\{[^}]*\}").unwrap(),
         "RCE-009", "OGNL/EL expression injection"),
        (Regex::new(r"\$\{jndi:").unwrap(),
         "RCE-010", "Log4Shell JNDI injection"),
        (Regex::new(r"(\{\{[^}]+\}\}|\{%[^%]+%\})").unwrap(),
         "RCE-011", "Server-side template injection"),
        (Regex::new(r"#\{T\s*\(").unwrap(),
         "RCE-012", "Spring SpEL injection"),
        (Regex::new(r"\b(require|__import__)\s*\(\s*['\x22](os|subprocess|socket)['\x22]\s*\)").unwrap(),
         "RCE-013", "Python/Ruby import exec"),
    ]
});

pub fn check(normalised: &str) -> Vec<Hit> {
    RULES.iter().filter_map(|(re, id, desc)| {
        if re.is_match(normalised) {
            Some(Hit { rule_id: id, description: desc, category: "rce", score: SCORE })
        } else {
            None
        }
    }).collect()
}
